### PR TITLE
[SPARK-10478][ML] Performance, organization, and style improvements for multi-layer perceptron

### PR DESCRIPTION
**Pull request by feynmanliang**
- Changes manual iteration loops into  `UFunc`s, vectorized, and broadcasted operations
- Refactors into more Scala idiomatic syntax (e.g. `while` and `for` -> `foreach` and `map`s, `System.arraycopy` -> `++`)
- Fixes various style issues (4-indent method args, wrap one-line ifs in `{}`)
- Adds comments and improves scaladocs (grammar, punctuation)

**Notes to Reviewers**
- Using `UFunc`s simplifies `ActivationFunction` code significantly at the cost of more complicated implementations for `crossEntropy`, `derivative`, etc... I'm not sure if the tradeoff here is worth it
- There is a slight performance hit in `SoftmaxFunction.eval` since we add an additional iteration over each column of `x` (previously computing exp(x - maxVal) and accumulating a sum were done at the same time whereas now the exp(x - maxVal) computation is done in one loop and the sum is computed after), but I feel that this is acceptable given the significant reduction in complexity. Thoughts?

CC @avulanov @mengxr

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala [hunk 1]
```diff
@@ -573,47 +582,27 @@ private[ml] class FeedForwardModel private(
...
-    // TODO: extract roll
-    for (i <- 0 until grads.length) {
-      val gradArray = grads(i)
-      var k = 0
-      while (k < gradArray.length) {
-        cumGradientArray(offset + k) += gradArray(k)
-        k += 1
-      }
-      offset += gradArray.length
+    grads.flatten.zipWithIndex.foreach { case (newGrad, i) =>
```
Comment (avulanov) [inline mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala]: Flatten might be expensive for array of large arrays, is not it?

File: mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala [hunk 2]
```diff
@@ -573,47 +582,27 @@ private[ml] class FeedForwardModel private(
...
-    }
-    val array = new Array[Double](size)
-    var offset = 0
-    for (i <- 0 until layerModels.length) {
-      val layerWeights = layerModels(i).weights().toArray
-      System.arraycopy(layerWeights, 0, array, offset, layerWeights.length)
-      offset += layerWeights.length
-    }
-    Vectors.dense(array)
+    Vectors.dense(layerModels.flatMap(_.weights().toArray))
```
Comment (avulanov) [inline mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala]: Ditto

**Comment by avulanov** [comment]
@feynmanliang Thank you for reviewing the code! I made one pass. It seems that UFunc simplifies it a lot. However I am not sure about `.flatten` and `.flatMap` on array of large arrays. We need to perform performance comparison. Could you run the benchmark from https://github.com/avulanov/ann-benchmark before and after refactoring to see the difference?

File: mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala (hunk 2 above)
Comment (brkyvz) [inline mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala]: I imagine System.arraycopy would be much faster and efficient here

**Comment by feynmanliang** [comment]
@avulanov The benchmarking code is written against a WIP implementation; I sent you a PR for bringing it up to date.

LBFGS is taking significantly long time on my machine:
image

I've removed the `flatten/flatMap` changes from this PR and will save them for when I have more time to properly perf test.

**Comment by avulanov** [comment]
@feynmanliang I suggest using native BLAS for testing. It worth checking the impact of using UFunc as well.

File: mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala [hunk 3]
```diff
@@ -260,127 +263,126 @@ private[ann] trait ActivationFunction extends Serializable {
...
-        y(i, j) /= sum
-        i += 1
-      }
-      j += 1
+    (0 until x.cols).foreach { j =>
+      // subtract max value to prevent overflow during exp
+      // does not affect correctness since we normalize right after
+      val maxVal = Bmax(x(::, j))
+      y(::, j) := breeze.numerics.exp(x(::, j) - maxVal)
+      y(::, j) :/= Bsum(y(::, j))
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala]: @feynmanliang Could you run some micro-benchmark on this function? I think this is the only place that might cause performance issues.

Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala]: @mengxr [Local benchmarks here](https://gist.github.com/feynmanliang/bc64b82a1258c4e86b9a). Performance is more or less equivalent.

**Comment by feynmanliang** [comment]
@mengxr added benchmarks, can you make another pass when you have a chance

**Pull request closed without merging**